### ISSUE TITLE
Providing an `init-rest` clause inside a class body breaks propagation of unused by-name initialization arguments to superclass

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#2395]
This program runs successfully:

```racket
#lang racket

(define foo%
  (class object%
    (init a)
    (super-new)))

(define bar%
  (class foo%
    (super-new)))

(new bar% [a 1])
```

…but this program fails with an error:

```racket
#lang racket

(define foo%
  (class object%
    (init a)
    (super-new)))

(define bar%
  (class foo%
    (init-rest args)
    (super-new)))

(new bar% [a 1])
```
```
instantiate: unused initialization arguments
  unused arguments:
   [a 1]
  instantiated class name: bar%
```

**Comment by lexi-lambda** [comment]
Looking into the code, this appears to be a “feature”:

https://github.com/racket/racket/blob/master/racket/collects/racket/private/class-internal.rkt#L3588-L3591

However, this is not documented, as far as I can tell, and I also don’t think it’s very helpful. My interpretation of `init-rest` is that, since by-name arguments are normally automatically propagated to the superclass, but by-position arguments are not, `init-rest` could be used to create a mixin that propagates *all* initialization arguments to the superclass, whether by-name *or* by-position:

```racket
(define (well-behaved-mixin %)
  (class %
    (init-rest args)
    (apply super-make-object args)))
```

However, this pattern doesn’t work, since `init-rest` disables automatic propagation of by-name arguments.

Is there something I’m missing? If not, is there a reason this behavior cannot be changed?

**Comment by mflatt** [comment]
I agree that this behavior isn't documented, at least as far as I can tell. It looks like the behavior of `(init-rest)` with no variable is also not documented.

I remember why `(init-rest)` behaves the way it does. It cuts off the propagation of by-position arguments. Then a superclass's initialization arguments cannot be supplied as by-position arguments when instantiating the derived class. Of all of a superclass's by-name argument are provided with `super-new`, then none of them can be supplied externally when instantiating the subclass.

The "feature" you found seems to have a dual effect: it cuts off the propagation of by-name arguments. I don't recall whether that was really the intent, but it seems possible that some program uses `init-rest` with an identifier to cut off propagation of by-name arguments.

**Comment by lexi-lambda** [comment]
I can accept that the behavior as-implemented cannot be changed now, but I suppose this just reveals I am asking an XY problem. I really just want a way to write a mixin that can be applied to a class with by-position and/or by-name arguments and forward all of its arguments to its superclass, as my `well-behaved-mixin` attempted to do in my previous comment. Is there a way to do that in the class system?

If I know ahead of time that only by-name or only by-position arguments will be relevant, I can make things work. If only by-name arguments are used in the superclass, then I can write

```racket
(define (well-behaved-mixin/by-name %)
  (class %
    (super-new)))
```

and by-name arguments will be propagated automatically. If only by-position arguments are used, I can instead write

```racket
(define (well-behaved-mixin/by-position %)
  (class %
    (init-rest args)
    (apply super-make-object args)))
```

but I have no idea how to get both.

**Comment by rfindler** [comment]
I've lost track of a lot of details over time, but this seems to do what you want? Did you need to inspect them?

```
#lang racket

(define (m x) (class x (super-new)))

(new (m (class object%
          (init a)
          (printf "a ~a\n" a)
          (super-new)))
     [a 1])

(make-object (m (class object%
                  (init-rest stuff)
                  (printf "stuff ~a\n" stuff)
                  (super-new)))
  1 2 3)
```

**Comment by lexi-lambda** [comment]
What you provide does seem to work, @rfindler, but I admit I’m a little surprised! Upon a closer reading of the documentation, I think I understand, but I found it confusing. It says:

> Unused (**by-name**) arguments are to be propagated to the superclass, as described in [Creating Objects](http://docs.racket-lang.org/reference/objcreation.html).

Emphasis mine. The linked section states:

> **By-name** initialization arguments to a class that have no matching initialization variable are implicitly added as by-name arguments to a `super-make-object`, `super-new`, or `super-instantiate` invocation, after the explicit arguments. If multiple initialization arguments are provided for the same name, the first (if any) is used, and the unused arguments are propagated to the superclass. (Note that converted by-position arguments are always placed before explicit by-name arguments.) The initialization procedure for the `object%` class accepts zero initialization arguments; if it receives any **by-name** initialization arguments, then `exn:fail:object` exception is raised.

Again, emphasis mine. The documentation suggests that only unused *by-name* arguments are propagated to the superclass.

My guess is that these “by-name” qualifiers aren’t actually very meaningful, since by-position arguments are immediately converted to by-name arguments, anyway:

> Arguments provided by position are converted into by-name arguments using the order of `init` and `init-field` clauses and the order of variables within each clause.

In that case, the difference between by-position and by-name arguments seems like it ought to be largely irrelevant, since the documentation seems to imply that by-position arguments are converted as soon as initialization is invoked. Except that theory doesn’t actually make sense, since if it were true, the result of this program wouldn’t make any sense:

```racket
(new (class object%
       (init-rest args)
       (super-new)
       (printf "args: ~a\n" args))
     [a 1])
```
```
instantiate: unused initialization arguments
  unused arguments:
   [a 1]
```

So by-position arguments are clearly not always converted to by-name arguments immediately upon initialization, but it isn’t specified when that happens, and since the difference between by-name and by-position arguments is made relevant by the documentation specifying that some things only apply to by-name arguments, it seems important to explain the details behind that process.

**Comment by rfindler** [comment]
Sure.
